org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java

```diff
@@ -579,14 +579,29 @@ private void addResourceDirs(IClasspathDescriptor classpath, IProject project, M
   }
 
   private void addResourceFolder(IClasspathDescriptor classpath, IPath resourceFolder, IPath outputPath,
-      boolean addTestFlag) {
+      boolean addTestFlag, Resource resource) {
     log.info("Adding resource folder " + resourceFolder);
-    IClasspathEntryDescriptor descriptor = classpath.addSourceEntry(resourceFolder, outputPath, DEFAULT_INCLUSIONS,
-        new IPath[] {IPath.fromOSString("**")}, false /*optional*/);
+    IClasspathEntryDescriptor descriptor = classpath.addSourceEntry(resourceFolder, outputPath, 
+      toIPathList(resource.getIncludes(), null),
+      toIPathList(resource.getExcludes(), "**/*.java"), false /*optional*/);
     descriptor.setClasspathAttribute(IClasspathManager.TEST_ATTRIBUTE, addTestFlag ? "true" : null);
     descriptor.setClasspathAttribute(IClasspathAttribute.OPTIONAL, "true"); //$NON-NLS-1$
   }
 
+  private IPath[] toIPathList(final List<String> fileNames, final String defaultPattern) {
+    if (fileNames == null) {
+      return defaultPattern != null ? new IPath[] {IPath.fromOSString(defaultPattern)} : DEFAULT_INCLUSIONS;
+    }
+    final List<IPath> retList = new ArrayList<>();
+    for (final String files : fileNames) {
+      retList.add(IPath.fromOSString(files));
+    }
+    if (defaultPattern != null) {
+      retList.add(IPath.fromOSString(defaultPattern));
+    }
```

Comment:
If a default, pattern is specified I don't think it is supposed to be added in any case.
Please provide a reference if I'm mistaken.

Reply:
toIPathList(resource.getIncludes(), null) ==> defaultPattern  is null

Reply:
Yes, but I meant the opposite, `defaultPattern` shall not be added at all if filenames is not null. Otherwise the defaults are added all the time, which is not the purpose of defaults.
```suggestion
```